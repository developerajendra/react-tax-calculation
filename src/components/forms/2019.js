import React,{useState} from 'react';

const INPUT_LIST = [
    'totalSalary',
    'extraAllowence',
    'HRA',
    'eightyC',
    'internet',
    'books',
    'foodWallet',
    'training',
    'LTA',
    'driverSalary',
    'fuel',
    'eightyD'
];

/**
 * Calculating the tax
 * @param {*} state 
 */
const calculateTax = (state) =>{
    let {totalSalary,extraAllowence, HRA, eightyC, internet,books, foodWallet, training,LTA,driverSalary,fuel,eightyD} = state;
    let range1 = 500000, range2 = 1000000;

    if(totalSalary<=range1){
        return{
            totalSalary,
            tax:0
        }
    }

    let total = (totalSalary+extraAllowence)-(HRA+eightyC+internet+books+foodWallet+training+LTA+driverSalary+fuel+eightyD);

    let calculate = (total)=>{
        let tax = 0;

        let calculateAmount = (slab,range)=>{
            let taxAmount = total-range;
            tax += (taxAmount)*slab/100;
            total = total-taxAmount;
        };

        //In case of salary is greater the 1000000
        if(total>range2){
           calculateAmount(30, range2)
        }

         //In case of salary is greater the 500000
        if(total>range1){
            calculateAmount(20, range1)
        }

        //In case of salary is less the 500000
        if(total<=range1){
            total = total - 250000;
            tax += (total)*5/100;
        }

        return tax;
    }
    calculate(total);
    
    return{
        totalSalary:totalSalary-calculate(total),
        tax:calculate(total)
    }
};


/**
 * Generating the input element
 * @param {*} list 
 */
let input = (state, setstate, list)=>{
    return list && list.map(data=>{
        return(
            <input type='number' value={state[data]}  placeholder={`${data}`} onChange={(e)=>setstate({...state,[data]:e.target.value*1})}></input>
        );
    })
};


/**
 * Generating the objects for default state
 * @param {*} list 
 */

let defaultState = (list)=>{
    let obj = {};
     list.map(data=>{
        obj = {
            ...obj,
            [data]:''
        }
    });
    return obj;
};

export const Forms2019 = () => {
    const [state, setstate] = useState({...defaultState(INPUT_LIST)});
    let _calculateTax = calculateTax(state);
    let _input = input(state, setstate, INPUT_LIST);

    return  (
            <div className="form-wrapper">
                <form>
                    {_input}
                </form>
                <div className="summary">
                    <h1>Summary</h1>
                    Salary: {(state.totalSalary*1)+(state.extraAllowence*1)}<br/>
                    Total after Tax: {_calculateTax.totalSalary} <br/>
                    Tax: {_calculateTax.tax}
                </div>
            </div>
    )
};
