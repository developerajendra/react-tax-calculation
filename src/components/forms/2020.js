import React,{Component, useState} from 'react';

let calculateTotal = (salary)=>{
    if(!salary) return{};
    let total, tax = null;
    if(salary<=500000){
        total = ' â‚¹' + salary;
        tax = 0;
    }else{
        let inc = 0;
        let _totalTax = 0;
        let calculate = (taxableSalary)=>{
            inc += 5;
            taxableSalary -= 250000;
            if(taxableSalary>250000){
                if(inc == 30){
                    _totalTax += taxableSalary*inc/100;
                }else{
                    _totalTax += 250000*inc/100;
                    calculate(taxableSalary);
                }
            }else{
                _totalTax += taxableSalary*inc/100;
            }
        }
        calculate(salary);
        var withCessTax = _totalTax + (_totalTax*4/100);
        total = salary - withCessTax;
    }
    return {
        total,
        tax:withCessTax
    }
};

export const Forms2020 = ()=> {
    const [state, setstate] = useState({totalSalary:''});
    let _calculateTotal = calculateTotal(state.totalSalary);

        return  (
            <div className="form-wrapper">
                <form>
                    <input type='number' value={state.totalSalary}  placeholder='Total Salary' onChange={(e)=>setstate({...state,totalSalary:e.target.value})}></input>
                </form>
                <div className="summary">
                    <h1>Summary</h1>
                    Salary: {state.totalSalary}<br/>
                    Total after Tax: {_calculateTotal.total} <br/>
                    Tax: {_calculateTotal.tax}
                </div>
            </div>
        )
}