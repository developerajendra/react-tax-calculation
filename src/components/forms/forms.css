.form-wrapper{
    display: flex;
    justify-content: space-around;
    width: 100%;
    padding-top: 62px;
}
.form-wrapper h1{
    margin: 0;
}
form {
    width: 50%;
}

input{
    padding: 20px;
    width: 100%;
    margin: 1px 0;
    border:none;
    box-sizing: border-box;
}
button{
    padding: 10px;
    border:none;
    background: #ccc;
    color:#000;
    width: 100%;

}
.title{
    text-transform: uppercase;
    font-size: 15px;
    padding: 20px;
    color: burlywood;
}
