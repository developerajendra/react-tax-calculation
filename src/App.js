import React from 'react';
import logo from './logo.svg';
import './App.css';

import {Forms} from './components/forms/forms';
import {UseStatePractice} from './components/useState/useState01'

function App() {
  return (
    <div className="App">
      {/* <UseStatePractice /> */}
      <header className="App-header">
        <Forms text='Tax Calculation'></Forms>
      </header>
    </div>
  );
} 

export default App;
