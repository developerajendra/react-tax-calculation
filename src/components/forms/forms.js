import React,{Component, Fragment, useState, useEffect} from 'react';
import Select from '../select/select';

import {Forms2020} from './2020';
import {Forms2019} from './2019';
 
export const Forms = (props)=>{
    let {text} = props;
    const [selectedYear, onStateChange] = useState(2020);
    const getFormState = (value)=>{
        onStateChange(value);
    };

    return  (
        <Fragment>
            <span className="title">{text}</span> <Select getFormState={getFormState}></Select>
           { selectedYear == '2020' ? 
            <Forms2020></Forms2020> :
            <Forms2019></Forms2019>}
        </Fragment>
    );
};