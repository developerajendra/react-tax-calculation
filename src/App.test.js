import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';

import Adapter from 'enzyme-adapter-react-16';

import App from './App';
import {Forms} from './components/forms/forms';
// import {add} from './sum';

 

test('render correctly app component', () => {  
  const appComponent = renderer.create(<App />).toJSON();
  expect(appComponent).toMatchSnapshot();
});

test('renders without creashing',()=>{
  const div = document.createElement('div');
  ReactDOM.render(<App/>, div);
});

test('rener correctly form component',()=>{
  const formComponent = renderer.create(<Forms/>).toJSON();
  expect(formComponent).toMatchSnapshot();
});
 
// test('adding the two numbers', ()=>{
//   expect(add(1,2)).toBe(3);
// });

// it('check form component should have title props',()=>{
//   const props = {
//     title:'test'
//   }, FormsComponent = mount(<Forms/>);
//   expect((FormsComponent).prop('title')).toEqual('test');
// });

 