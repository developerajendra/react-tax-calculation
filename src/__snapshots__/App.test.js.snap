// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly app component 1`] = `
<div
  className="App"
>
  <div>
    <button>
      click me
    </button>
  </div>
  <header
    className="App-header"
  >
    <span
      className="title"
    >
      Tax Calculation
    </span>
     
    <select
      id="select"
      name="yarSelect"
      onChange={[Function]}
    >
      <option
        value="2020"
      >
        2020
      </option>
      <option
        value="2019"
      >
        2019
      </option>
    </select>
    <form>
      <h1>
        2020
      </h1>
      <input
        placeholder="Total Salary"
        type="number"
      />
      <input
        placeholder="Extra Allowence"
        type="number"
      />
      <input
        placeholder="HRA"
        type="number"
      />
      <input
        placeholder="80C"
        type="number"
      />
      <input
        placeholder="Internet"
        type="number"
      />
      <input
        placeholder="Books"
        type="number"
      />
      <input
        placeholder="Food Wallet"
        type="number"
      />
      <input
        placeholder="Training & Certification"
        type="number"
      />
      <input
        placeholder="LTA"
        type="number"
      />
      <input
        placeholder="Driver Salary"
        type="number"
      />
      <input
        placeholder="Fuel"
        type="number"
      />
      <input
        placeholder="80D"
        type="number"
      />
      <button>
        Get Summary
      </button>
    </form>
  </header>
</div>
`;

exports[`rener correctly form component 1`] = `
Array [
  <span
    className="title"
  >
    title
  </span>,
  " ",
  <select
    id="select"
    name="yarSelect"
    onChange={[Function]}
  >
    <option
      value="2020"
    >
      2020
    </option>
    <option
      value="2019"
    >
      2019
    </option>
  </select>,
  <form>
    <h1>
      2020
    </h1>
    <input
      placeholder="Total Salary"
      type="number"
    />
    <input
      placeholder="Extra Allowence"
      type="number"
    />
    <input
      placeholder="HRA"
      type="number"
    />
    <input
      placeholder="80C"
      type="number"
    />
    <input
      placeholder="Internet"
      type="number"
    />
    <input
      placeholder="Books"
      type="number"
    />
    <input
      placeholder="Food Wallet"
      type="number"
    />
    <input
      placeholder="Training & Certification"
      type="number"
    />
    <input
      placeholder="LTA"
      type="number"
    />
    <input
      placeholder="Driver Salary"
      type="number"
    />
    <input
      placeholder="Fuel"
      type="number"
    />
    <input
      placeholder="80D"
      type="number"
    />
    <button>
      Get Summary
    </button>
  </form>,
]
`;
